Fix cacorrect multithreading

jenshannoschwalm opened the pull request:
Make sure all internal buffers are aligned-64 to allow vectorisation avoiding overflow from one to the next buffer.
Ideas:
1. Assume the OpenMP is correct
2. Compiler does some vectorisation
3. Make sure all internal buffers are strictly aligned to 64

@sh1llo try to get into your issue (fearing it might not be a compiler but a design problem in the algo), could you try this

EDIT: will keep working on this until fine ...

Fixes #15600

jenshannoschwalm commented:
force pushed some workup, code readability, rounding fixes and declaring some parameters shared

I looked up file history here and spotted #15056 and #15116, i think the fix by @MStraeten didn't fix an apple but a clang issue.

I would like
1. @sh1llo to test this pr as is and
2. @MStraeten or @zisoft to test this PR plus remove the test for apple in `#if defined(_OPENMP) && !defined(__APPLE__)` in line 379

MStraeten commented:
unfortunately this PR dosen't fix #15116, so disabling that parallelization at least for Apple M1 can't be dropped ...

jenshannoschwalm commented:
Thanks Martin. Could you test with that passage being

```
#if defined(_OPENMP)
#pragma omp parallel
#endif

```

zisoft commented:
So here are my results (Intel Mac, AMD Radeon 5500):

**current master without this PR**
 apply raw chromatic abberation --> ok

uncomment `#if defined(_OPENMP) && !defined(__APPLE__)`
apply raw chromatic abberation --> dt hangs (working...)


**current master with this PR**
apply raw chromatic abberation --> ok

uncomment `#if defined(_OPENMP) && !defined(__APPLE__)`
apply raw chromatic abberation --> dt hangs (working...)

So unfortunately no change for Apple

jenshannoschwalm commented:
Thanks a lot for testing. It seems we have to test for gcc & openmp here.

Unfortunately I have no idea why this is so. Do we hit some building issue? Is  this a clang problem? Does our code not follow openmp specs?

jenshannoschwalm commented:
Latest commit reverted some OpenMP pragmas, this make gcc running system here ~5% slower but hopefully allows running with threads on clang builds.

@MStraeten and @zisoft

zisoft commented:
Tim Cook will hire you :-)

Works!!
And as far I can tell it is much faster now.

MStraeten commented:
fixes #15116 on arm64

jenshannoschwalm commented:
> Tim Cook will hire you :-)

I don't care about **that** but i live very close in Hamm so i would accept a coffee afetr a bike ride and see how you do things in darktable on a mac :-)

jenshannoschwalm commented:
> And as far I can tell it is much faster now.

It should be, it's using multiple threads now :-)

jenshannoschwalm commented:
@TurboGit i think ready to get merged now.

zisoft commented:
> i live very close in Hamm so i would accept a coffee afetr a bike ride and see how you do things in darktable on a mac :-)

very much appreciated!

sh1llo commented:
Tested with this pr, fixes the issue for me. Just to make sure, I checked my older edits that were freezing, and they also correctly open now.
+1 from me, and thanks!

dterrahe commented on src/iop/cacorrect.c:
```diff
@@ -440,7 +443,7 @@ void process(
       for(int top = -border; top < height; top += ts - border2)
```
{

Nothing to do with your changes, but I'm surprised Pascal didn't ask for it :-)

jenshannoschwalm replied:
indeed, makes that huge loop better to review

TurboGit replied:
I hadn't reviews yet :)

dterrahe commented on src/iop/cacorrect.c:
```diff
@@ -340,93 +343,93 @@ void process(
     rgb[0] = (float(*))data;
-    rgb[1] = (float(*))(data + 1 * sizeof(float) * ts * ts + 1 * 64);
-    rgb[2] = (float(*))(data + 2 * sizeof(float) * ts * ts + 2 * 64);
+    rgb[1] = (float(*))(data + tilebuf_size);
```
You can get rid of the explicit casts and sets of parentheses now, which makes this piece of code a bit more readable.

jenshannoschwalm replied:
rigth again

jenshannoschwalm commented:
force pushed squashed suggested modifications

dterrahe commented:
Just for my understanding/education (and maybe a bit of documentation, in case someone in future tries to claw back those 5% performance lost and inadvertently reintroduces the bug!) removing the "default(none)" pragma and explicit sharing declarations reverts all variables to _shared_.

Presumably this doesn't impact any of the consts (?) so that leaves _darktable_ (which is read in infrequently executed `dt_print` only) and _processpasstwo_. The latter doesn't seem to be referenced in an inner loop, so is the 5% hit surprising?

jenshannoschwalm commented:
Nope, I would say and I don't see a way atm with other pragmas. We might try with simd and aligning hints though.

My major lesson for now - and I should have spotted it earlier - if there is something wrong about apple, it could be clang.

Maybe we should also do integration tests with clang?

TurboGit reviewed (approved):
Thanks!

TurboGit commented:
I'll have to analyze the regression test:

```
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 43.13434
      Avg dE                   : 0.65424
      Std dE                   : 3.53237
      ----------------------------------
      Pixels below avg + 0 std : 93.87 %
      Pixels below avg + 1 std : 96.61 %
      Pixels below avg + 3 std : 97.71 %
      Pixels below avg + 6 std : 98.80 %
      Pixels below avg + 9 std : 99.94 %
      ----------------------------------
      Pixels above tolerance   : 4.03 %
 
  FAILS: image visually changed
         see diff.png for visual difference
         (308908 pixels changed)

```

TurboGit commented:
Hum... the diff is very suspicious as the right border is very different:

![image](https://github.com/darktable-org/darktable/assets/467069/e5be59ce-55d7-4fda-bb0b-4bf1ef35e6ea)

But the right border is a uniform color, and well it is indeed broken, see the new output:

![image](https://github.com/darktable-org/darktable/assets/467069/a6c39156-285d-4824-9428-a3e7136a2179)

So after all not ready for integration.

TurboGit reviewed (changes requested):
Something broken...

jenshannoschwalm commented:
Will check and look if there was a problem or if this is a new problem.

MStraeten commented:
can't reproduce that right border issue on my mac but have reproduced it with a windows build

TurboGit commented:
@jenshannoschwalm : For the record, the first commit (2c9117e58ef6bce0f85c59bc98afd6d3b5e7ae1a) is introducing this regression.

zisoft commented:
Same here, not reproducible on my Mac, but on Windows.
The green border appears when _avoid colorshift_ is checked.

Another thing I have noted: On Mac there is absolutely no visible difference when activating/deactivating the module.
No change in waveform or vectorscope.

Tested with mire1.cr2

jenshannoschwalm commented:
Thanks for the first-commit suggestion. I looked at it again and there is indeed a very stupid bug when calculating the number of tiles. Will fix asap

jenshannoschwalm commented:
Could you all test again, i think i understand the things i did wrong. Made sure about aligning and fixed (also renamed vars for readbility) wrongdoing for number of tiles to be calculated resulting in the border issue.

@TurboGit could you please check the integration test on your side? If everything goes right and no more requests after reviewing i would propose to
1. squash in 1 commit so everything is in one place
2. correct the commit message

TurboGit commented:
@jenshannoschwalm : The test pass and is 100% identical to the expected output:

```
Test 0084-cacorrect
      Image mire1.cr2
      CPU & GPU version differ by 18288 pixels
      CPU vs. GPU report :
      ----------------------------------
      Max dE                   : 1.61430
      Avg dE                   : 0.00296
      Std dE                   : 0.04009
      ----------------------------------
      Pixels below avg + 0 std : 99.34 %
      Pixels below avg + 1 std : 99.35 %
      Pixels below avg + 3 std : 99.37 %
      Pixels below avg + 6 std : 99.46 %
      Pixels below avg + 9 std : 99.55 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
      Expected CPU vs. current CPU report :
      ----------------------------------
      Max dE                   : 0.00000
      Avg dE                   : 0.00000
      Std dE                   : 0.00000
      ----------------------------------
      Pixels below avg + 0 std : 100.00 %
      Pixels below avg + 1 std : 100.00 %
      Pixels below avg + 3 std : 100.00 %
      Pixels below avg + 6 std : 100.00 %
      Pixels below avg + 9 std : 100.00 %
      ----------------------------------
      Pixels above tolerance   : 0.00 %
 
  OK
```

jenshannoschwalm commented:
Good. Will do the squashing later tonight.

jenshannoschwalm commented:
Added another macro borderh for code readability, no algo stuff changed. Ready to go

TurboGit reviewed (approved):
All good now, thanks!

jenshannoschwalm commented:
@ralfbrown could you have look too? Anything compiler specific, clang? Or some aligning stuff for osx?

ralfbrown commented:
I'm not aware of any compiler-specific issues, but I also don't use clang.  Cacheline-alignment shouldn't have any impact other than performance (due to cachelines getting bounced between the two cores processing the pixels before/after the boundary a cacheline straddles).

I'll take a look, but it's a hefty chunk of code so I'll need to dedicate a block of time.  Probably over the weekend.

jenshannoschwalm commented:
> I'll take a look, but it's a hefty chunk of code so I'll need to dedicate a block of time. Probably over the weekend.

:-) It really is hefty.
